### DISCUSSION TITLE
Two packages in one Distribution Package using pyproject.toml only

### DISCUSSION THREAD

**Discussion by buhtz** [pypa/setuptools#3725] [Q&A] [answered]
This was born from https://github.com/pypa/setuptools/discussions/3566.
The difference is that now I ask if this is possible to achieve with using `pyproject.toml` file only **without** a `setup.cfg` and/or `setup.py` file.

The goals are
 - Having a [Distribution Package](https://packaging.python.org/en/latest/glossary/#term-Distribution-Package) named `bit_demo` or `bitdemo`
 - that should contain two [Import Package](https://packaging.python.org/en/latest/glossary/#term-Import-Package) named `bitcli` and `bitgui`.
 - Using the "src-Layout" for the project

I want to create a demonstrator for a project like this. The repo can be found at [Codeberg.org](https://codeberg.org/buhtz/bit_demo).

The structure of folders and files are currently like this

```
bit_demo
├── pyproject.toml
├── src
│   ├── bitcli
│   │   ├── __init__.py
│   │   └── __main__.py
│   └── bitgui
│       ├── __init__.py
│       └── __main__.py
└── tests
    ├── integration
    └── unit
```

The `pyproject.toml`
```
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[project]
name = "bitdemo"
version = "0.0.1"

[tool.setuptools.package-dir]
package-dir =
    bitdemo = src/bitdemo
    bitgui = src/bitgui
```

Something is wrong with the `package-dir =` line because of that error when doing `python3 -m pip install -e .`

`pip._vendor.tomli.TOMLDecodeError: Invalid value (at line 10, column 14)`

- My primary problem is that error message.
- Is it even possible to have two Import Packages in one Distribution Package using only a `pyproject.toml`?
- From the [setuptools-docu](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use) it is not clear for me if I need a `[build-system]` section when not having a `setup.py` and/or `setup.cfg`
- If I need a `[build-system]` section I'm not clear about the options in there. I find examples with `setuptools` but also with `hatchin`.
- The project name (`bitdemo`) in `pyproject.toml` is different from the repo's root folder name `bit_demo`. Does this conflicts?

Keep in mind that you can also look [into the original files in my Codeberg.org repo](https://codeberg.org/buhtz/bit_demo).

**Comment by buhtz** [comment]
I made stupid mistakes and typos. It works now.
I modified the "package-dir" section.

```
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"

[project]
name = "bitdemo"
version = "0.0.1"

[tool.setuptools.package-dir]
bitcli = "src/bitcli"
bitgui = "src/bitgui"
```

Now it works as expected.
```
Python 3.9.2 (default, Feb 28 2021, 17:03:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import bitdemo
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'bitdemo'
>>> import bitgui
>>> import bitcli
```
`bitdemo` fails because it is the Distribution Package but nothing that can be imported.

But some of the questions at the end of my initial post are still open.

**Comment by abravalheri** [comment, marked as answer]
Hi @buhtz , as a golden rule I always try to make things as simpele/straight forward as possible and only deviate from that if I have a good reason or if it is not possible to do otherwise.

Please find my attempt to answer your questions below:

> it is not clear for me if I need a [build-system] section when not having a setup.py and/or setup.cfg

 `[build-system]` is always recommend. You should always have one...

> Is it even possible to have two Import Packages in one Distribution Package using only a pyproject.toml?

It is possible. If you know what you are doing , that is completely fine.

However this is generally discouraged. It is less confusing to have a 1:1 mapping. Moreover 1:1 mapping avoids name clashing in PyPI (if other dev decides to upload a different distribution using one of your packages names that can cause problems/confusion for your users).

> If I need a [build-system] section I'm not clear about the options in there. I find examples with setuptools but also with hatchin.

If you are using `setuptools` and following `setuptools` documentation you should set  `build-backend = "setuptools.build_meta"` inside of `[build-system]`. `hatchling` is a different tool and will build projects differently.

[This link](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#basic-use) has a very simple example for this small section of the file.

> The project name (bitdemo) in pyproject.toml is different from the repo's root folder name bit_demo. Does this conflicts?

This is possible, but generally discouraged unless you have a good reason. Apart from using `-` in distribution names and `_` in folders/files, it is simpler to keep things 1:1. I find that it avoids confusion and prevents name clashes in PyPI.

**Reply by buhtz** [reply]
Thank you very much for your reply and answers.
I don't intent to upload the distribution package to PyPi. At this point I know what I'm doing. In my case there is much more value in having a 1:2 distro:import then splitting it into two repos.
